system/storage/bucket.toit

```diff
@@ -15,92 +15,93 @@
 
 import .storage show StorageServiceProvider
 
-import binary show LITTLE_ENDIAN
-import encoding.tison
+import encoding.base64
+import system.assets
 import system.services show ServiceResource
 import uuid
 
 abstract class BucketResource extends ServiceResource:
+  // We keep a cache of computed ids around. The ids encode
+  // the root name of the bucket, so it is natural to keep
+  // these per bucket. This also simplifies cleaning them
+  // up as the cache simply goes away when the bucket
+  // resource is closed.
+  ids_ ::= {:}
+
   constructor provider/StorageServiceProvider client/int:
     super provider client
 
+  abstract root -> string
   abstract get key/string -> ByteArray?
   abstract set key/string value/ByteArray -> none
   abstract remove key/string -> none
 
   on_closed -> none:
     // Do nothing.
 
+  compute_id_ key/string -> string:
+    return ids_.get key --init=:
+      id := uuid.uuid5 root key
+      encoded := base64.encode id.to_byte_array
+      // Keys used in nvs must be 15 bytes or less, so we
+      // pick the first 12 encoded bytes which correspond
+      // to the first 9 bytes of the 16 uuid bytes.
+      encoded[..12]
+
 class FlashBucketResource extends BucketResource:
   static group ::= flash_kv_init_ "nvs" "toit" false
   root/string
-  paths ::= {:}
-
   constructor provider/StorageServiceProvider client/int .root:
     super provider client
 
   get key/string -> ByteArray?:
-    return flash_kv_read_bytes_ group (compute_path_ key)
+    return flash_kv_read_bytes_ group (compute_id_ key)
 
   set key/string value/ByteArray -> none:
-    flash_kv_write_bytes_ group (compute_path_ key) value
+    flash_kv_write_bytes_ group (compute_id_ key) value
 
   remove key/string -> none:
-    flash_kv_delete_ group (compute_path_ key)
-
-  compute_path_ key/string -> string:
-    return paths.get key --init=: (uuid.uuid5 root key).stringify[..13]
+    flash_kv_delete_ group (compute_id_ key)
```

Comment:
In theory it could make sense to keep a delete counter and clear the `ids_` when we have done a lot of removals and have a big set.

Reply:
Yeah, or just cap the number of cached ids -- or the number of unused cached ids. For another PR :)